src/recordlinker/linkage/simple_mpi.py

```diff
@@ -0,0 +1,110 @@
+"""
+recordlinker.linkage.simple_mpi
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This module provides the data access functions to the MPI tables
+"""
+
+import typing
+
+from sqlalchemy import orm
+from sqlalchemy.sql import expression
+
+from recordlinker.linkage import models
+
+
+def get_block_data(
+    session: orm.Session, data: dict, algo_config: dict
+) -> list[models.Patient]:
+    """
+    Get all of the matching Patients for the given data using the provided
+    blocking keys defined in the algo_config.
+    """
+    # Create the base query
+    query = session.query(models.Patient)
+
+    # Build the join criteria, we are joining the Blocking Value table
+    # multiple times, once for each Blocking Key.  If a Patient record
+    # has a matching Blocking Value for all the Blocking Keys, then it
+    # is considered a match.
+    for idx, block in enumerate(algo_config["blocks"]):
+        key_name = block["value"].upper()
+        # Get the matching Blocking Key based on the value in the algo_config
+        assert hasattr(models.BlockingKey, key_name), f"Invalid Blocking Key: {block}"
+        key = models.BlockingKey[key_name]
+        # Get all the possible values from the data for this key
+        vals = [v for v in key.to_value(data)]
+        # Create a dynamic alias for the Blocking Value table using the index
+        # this is necessary since we are potentially joining the same table
+        # multiple times with different conditions
+        alias = orm.aliased(models.BlockingValue, name=f"bv{idx}")
+        # Add a join clause to the mpi_blocking_value table for each Blocking Key.
+        # This results in multiple joins to the same table, one for each Key, but
+        # with different joining conditions.
+        query = query.join(
```

Comment:
I can think of at least 3 ways of doing this.

1) Using a HAVING clause to check that the COUNT of all the BlockingKey matches equals the number of where conditions provided.  I think this would be the slowest, because we're having to do an aggregate on the data first.
2) Using EXISTS subqueries for each BlockingKey match.  This could be faster, but its a little harder to write.
3) Self Joins, which is what you see here in we are joining against the table multiple times.  I think this is the simplest to write and also comprehend, so I wanted to start with this one.

If this performs at least as fast as the existing schema, I'll likely stick with this strategy, however if its slower I will def try number 2.  This is a difficult thing to nail down, as it could be the second is faster with Postgresql and thrid is faster with Microsoft SQL Server.  There will be some trail and error here.  Compared to the [status quo](https://github.com/CDCgov/RecordLinker/blob/main/src/recordlinker/linkage/mpi.py#L240-L472), I think its a reasonable place to start.